src/builtins.ts

```diff
@@ -3522,20 +3522,39 @@ function builtin_i8x16(ctx: BuiltinContext): ExpressionRef {
   }
   var operands = ctx.operands;
   var bytes = new Uint8Array(16);
+  var vars  = new Array<ExpressionRef>(16);
+  var numVars = 0;
+
   for (let i = 0; i < 16; ++i) {
     let expr = compiler.compileExpression(operands[i], Type.i8, Constraints.CONV_IMPLICIT);
     let precomp = module.runExpression(expr, ExpressionRunnerFlags.PreserveSideeffects);
     if (precomp) {
       writeI8(getConstValueI32(precomp), bytes, i);
     } else {
-      compiler.error(
-        DiagnosticCode.Expression_must_be_a_compile_time_constant,
-        operands[i].range
-      );
+      vars[i] = expr;
+      numVars++;
     }
   }
   compiler.currentType = Type.v128;
-  return module.v128(bytes);
+  if (numVars == 0) {
+    // all constants
+    return module.v128(bytes);
+  } else {
+    let vec: ExpressionRef;
+    let fullVars = numVars == 16;
+    if (fullVars) {
+      // all variants
+      vec = module.unary(UnaryOp.SplatI8x16, vars[0]);
+    } else {
+      // mixed constants / variants
+      vec = module.v128(bytes);
+    }
+    for (let i = i32(fullVars); i < 16; i++) {
+      let expr = vars[i];
+      if (expr) vec = module.simd_replace(SIMDReplaceOp.ReplaceLaneI8x16, vec, <u8>i, expr);
```

Comment:
These should always be `true`, becomes `unreachable` on errors.

```suggestion
      vec = module.simd_replace(SIMDReplaceOp.ReplaceLaneI8x16, vec, <u8>i, vars[i]);
```